### PR TITLE
Alert user to failed concretizations

### PR DESCRIPTION
**Pull request by scheibelp**
Sometimes the solver phase of Spack succeeds but Spack quietly discards those results because the solver output is not found to satisfy the user's input specs. This is always an internal Spack error and we should tell the user this so they can submit a bug report.

This extracts the debugging parts of https://github.com/spack/spack/pull/41049 (which also highlight one of the cases where Spack fails silently) and makes it so that Spack always fails when it finds a discrepancy between solver output and requested specs (as a side effect, some logic that depended on the casual, informational aspect of `unsolved_specs` has been removed because this PR makes it fail before it could be examined in that manner).

### CODE REVIEW

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -776,6 +782,12 @@ def on_model(model):
             error_handler = ErrorHandler(best_model)
             error_handler.raise_if_errors()
 
+            if result._unsolved_specs:
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]:
```suggestion
            if result._unsolved_specs and setup.concretize_everything:
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/solver/asp.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: Thanks: looking into what this meant, I discovered my changes were not actually throwing at the right moment, so I need to correct this.

**Review by becker33 — DISMISSED** [review]

File: lib/spack/spack/test/concretize.py [hunk 2]
```diff
@@ -341,6 +341,7 @@ def test_different_compilers_get_different_flags(self):
         assert set(client.compiler_flags["fflags"]) == set(["-O0", "-g"])
         assert not set(cmake.compiler_flags["fflags"])
 
+    @pytest.mark.xfail(reason="Broken, needs to be fixed")
```
Comment (scheibelp) [inline lib/spack/spack/test/concretize.py:344]: Note that `Spec.concretize` and `concretize._concretize_specs_together_new` are inconsistent in how they interact with the result: the former pulls from `Result.answers` (which is populated even if the specs in the answer are not considered to satisfy) whereas the latter uses `Result.specs` (which omits solvers specs that do not satisfy). This PR makes the behavior more consistent (and in that sense, this test becomes a reproducer for what I originally had to create an env for in https://github.com/spack/spack/issues/38664).

File: lib/spack/spack/solver/asp.py [hunk 3]
```diff
@@ -468,8 +468,14 @@ def _compute_specs_from_answer_set(self):
 
             if candidate and candidate.satisfies(input_spec):
                 self._concrete_specs.append(answer[node])
-                self._concrete_specs_by_input[input_spec] = answer[node]
+                self._concrete_specs_by_input.append((input_spec, answer[node]))
```
Comment (scheibelp) [inline lib/spack/spack/solver/asp.py]: Notes:

* The only things that use `concrete_specs_by_input` call `update` with it, so it doesn't have to be a dictionary
* Without this, `test_concretize_partial_old_dag_hash_spec` fails (see https://github.com/spack/spack/actions/runs/7894899814/job/21546544138?pr=42655)
    * Overall, we shouldn't hash specs before calling `_finalize_concretization`
    * I don't think it makes sense to call `finalize_concretization` until we know the solver has succeeded
    * So to resolve this cycle, I removed the need for hashing by making this a `List[Tuple[Spec, Spec]]`, I considered adding an explicit typehint, but I'm concerned that would create a cyclic import (Specs need Solvers need Specs)

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: I'm revisiting this and undoing it because input/concrete specs should always be hashable. The test failure was more an issue with how the test was simulating a dependency without a package hash, so I have addressed that directly.

File: lib/spack/spack/cmd/solve.py [hunk 4]
```diff
@@ -126,12 +126,6 @@ def _process_result(result, show, required_format, kwargs):
                 sys.stdout.write(spec.tree(color=sys.stdout.isatty(), **kwargs))
         print()
 
-    if result.unsolved_specs and "solutions" in show:
-        tty.msg("Unsolved specs")
-        for spec in result.unsolved_specs:
-            print(spec)
-        print()
-
```
Comment (alalazo) [inline lib/spack/spack/cmd/solve.py]: Can we leave this one in? It is useful to check the flow of `when_possible` concretizations.

File: lib/spack/spack/test/concretize.py [hunk 5]
```diff
@@ -2093,7 +2094,25 @@ def test_result_specs_is_not_empty(self, specs):
         result, _, _ = solver.driver.solve(setup, specs, reuse=[])
 
         assert result.specs
-        assert not result.unsolved_specs
+
+    def test_unsolved_specs_raises_error(self, monkeypatch, mock_packages, config):
```
Comment (alalazo) [inline lib/spack/spack/test/concretize.py:2099]:
```suggestion
    @pytest.mark.regression("38664")
    def test_unsolved_specs_raises_error(self, monkeypatch, mock_packages, config):
```

File: lib/spack/spack/test/spec_semantics.py [hunk 6]
```diff
@@ -1303,7 +1303,7 @@ def test_spec_trim(mock_packages, config):
 def test_concretize_partial_old_dag_hash_spec(mock_packages, config):
     # create an "old" spec with no package hash
     bottom = Spec("dt-diamond-bottom").concretized()
-    delattr(bottom, "_package_hash")
+    bottom._package_hash = None
```
Comment (alalazo) [inline lib/spack/spack/test/spec_semantics.py]: Was this creating an issue? Or is the change unrelated?

File: lib/spack/spack/solver/asp.py [hunk 7]
```diff
@@ -805,6 +811,12 @@ def on_model(model):
             print("Statistics:")
             pprint.pprint(self.control.statistics)
 
+        if result.unsolved_specs and setup.concretize_everything:
+            raise UnsatisfiableSpecError(
+                "Internal Spack error: the solver completed but produced specs"
+                " that do not satisfy the request. Rerun with 'spack -d'"
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: Can we raise an `InternalConcretizerError`?

File: lib/spack/spack/solver/asp.py [hunk 8]
```diff
@@ -470,6 +470,12 @@ def _compute_specs_from_answer_set(self):
                 self._concrete_specs.append(answer[node])
                 self._concrete_specs_by_input[input_spec] = answer[node]
             else:
+                msg = f"Unsatisfied input spec: {str(input_spec)}"
+                if candidate:
+                    msg += f"\n\tCandidate spec: {str(candidate)}"
+                else:
+                    msg += "\n\t(No candidate specs from solver)"
+                tty.debug(msg)
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py]: (nitpick) I would avoid debugging messages here, as they might add a lot of entries in `when_possible` environments

File: lib/spack/spack/solver/asp.py [hunk 9]
```diff
@@ -805,6 +811,12 @@ def on_model(model):
             print("Statistics:")
             pprint.pprint(self.control.statistics)
 
+        if result.unsolved_specs and setup.concretize_everything:
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:826]: :+1: To this additional check. This configuration should never happen.

**Review by alalazo — CHANGES_REQUESTED** [review]
Only minor comments on debug messages. I agree on tests and adding an additional check at the end of `solve`.

**Review by becker33 — DISMISSED** [review]

File: lib/spack/spack/test/spec_semantics.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/test/spec_semantics.py]: Yes, if this change is removed, you get the failure in https://github.com/spack/spack/actions/runs/7894899814/job/21546544138

* This PR ends up hashing input specs as part of concretization (that wasn't done before)
  * Actually to be precise, it makes it so that specs are hashed via both of the two distinct concretization pathways we have (see https://github.com/spack/spack/pull/42655#discussion_r1488781515)
* This test provides an un-hashable input spec (`delattr` makes `set([spec])` fail)
* My first thought was that we should avoid hashing specs in general as part of this process (so I tried https://github.com/spack/spack/pull/42655#discussion_r1488810448)
* I then decided though that the specs hashed in this PR are concrete or user specs, which should always be OK to hash, and this test was violating the contract of the object, so I changed the simulation to match what would occur "naturally" if no package hash was available (the attribute should be present but `None`).

File: lib/spack/spack/solver/asp.py (hunk 8 above)
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: Fair enough: it was lazy of me to put this here, but in order to retain the information I'll need to update the structure of `_unsolved_specs` so that it can be retrieved at the later point (I want to show the user spec and associated candidate together).

File: lib/spack/spack/test/concretize.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/test/concretize.py:2099]: This doesn't actually fix 38664, but I might misunderstand what we use `mark.regression` for: I do like connecting this test to that conversation so I'll update this.

File: lib/spack/spack/test/spec_semantics.py (hunk 6 above)
Comment (alalazo) [reply lib/spack/spack/test/spec_semantics.py]: I'll try to investigate this. I would have expected it to not fail, since:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/spec.py#L4288

So I wonder which code path we are hitting. Note that this is tricky, because we don't set this attribute in `Spec.__init__` but in a very indirect way using:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/spec.py#L1895

where the attribute that gets written is also computed indirectly:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/hash_types.py#L31-L34

TL;DR Given the above, I don't think using `None` is testing the same situation we'd have with old specs. There it's likely that the `_package_hash` doesn't exist as an attribute (but I need to double check to be 100% sure).

File: lib/spack/spack/solver/asp.py (hunk 7 above)
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: `InternalConcretizerError` was originally specifically for cases where the solver could not find a solution (even with errors); I extended the exception hierarchy to add a `SolverError` as a subclass of `InternalConcretizerError` for this case.

File: lib/spack/spack/test/concretize.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/test/concretize.py:2099]: Done

File: lib/spack/spack/solver/asp.py (hunk 8 above)
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: `Result.unsolved_specs` is now a list of tuples (input spec, result spec)

* This info is now dumped with the raised exception (which also means the user doesn't have to rerun)
* https://github.com/spack/spack/pull/42655#discussion_r1499379674 is also updated to output the associated candidate specs as well
* It seems to me we could do an early termination of `solve_in_rounds` if any of the values of `unsolved_specs` set (not `None`) although I haven't done that here

File: lib/spack/spack/test/spec_semantics.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/test/spec_semantics.py]: I thought that because

```
delattr(spec, "_package_hash")
set([spec]) # <---- this fails
```

fails it meant that `delattr` was wrong, although after re-checking I think you're right that this should work (`Spec._finalize_concretization` claims some concrete Specs may not have this attr). I think the other approach is to update

```
def to_node_dict(...):
    if self._concrete and hash.package_hash and self._package_hash:
```

(i.e. the spot that generates the exception)

to

```
if self._concrete and hash.package_hash and hasattr(self, "_package_hash") and self._package_hash:
```

I'll say though that other methods of Spec may be inconsistent in what they expect of `_package_hash`, e.g. `Spec.package_hash` also seems to assume `_package_hash` would always be present.

Comment (alalazo) [reply lib/spack/spack/test/spec_semantics.py]: Not to be fixed in this PR, but as we are discussing it: are there any technical reason we use this convoluted method of setting a private attribute without declaring it in `__init__`?

**Review by alalazo — APPROVED** [review]

File: lib/spack/spack/solver/asp.py [hunk 10]
```diff
@@ -3447,8 +3470,21 @@ def __init__(self, msg):
 
 
 class InternalConcretizerError(spack.error.UnsatisfiableSpecError):
-    """
-    Subclass for new constructor signature for new concretizer
+    """Errors that indicate a bug in Spack."""
+
+    def __init__(self, msg):
+        super(spack.error.UnsatisfiableSpecError, self).__init__(msg)
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:3476]: I didn't notice this, but it's... weird. I'll follow up in another PR, but we are skipping one class in the MRO because we don't match its conctructor. I wonder why not inheriting from `spack.error.SpecError` then.

File: lib/spack/spack/test/spec_semantics.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/test/spec_semantics.py]: Not that I'm aware of. IMO setting `_package_hash = None`  in `__init__` would be better (but I could be overlooking logic that changes its behavior depending on whether the attr is present).
